src/parquet/column_chunk_reader.cc

```diff
@@ -0,0 +1,488 @@
+#include <seastar/parquet/column_chunk_reader.hh>
+#include "compression.hh"
+
+namespace parquet {
+
+seastar::future<std::optional<page>> page_reader::next_page() {
+    *_latest_header = format::PageHeader{}; // Thrift does not clear the structure by itself before writing to it.
+    return read_thrift_from_stream(_source, *_latest_header).then([this] (bool read) {
+        if (!read) {
+            return seastar::make_ready_future<std::optional<page>>();
+        }
+        if (_latest_header->compressed_page_size < 0) {
+            throw parquet_exception::corrupted_file(seastar::format(
+                    "Negative compressed_page_size in header: {}", *_latest_header));
+        }
+        size_t compressed_size = static_cast<uint32_t>(_latest_header->compressed_page_size);
+        return _source.peek(compressed_size).then(
+        [this, compressed_size] (std::basic_string_view<uint8_t> page_contents) {
+            if (page_contents.size() < compressed_size) {
+                throw parquet_exception::corrupted_file(seastar::format(
+                        "Unexpected end of column chunk while reading compressed page contents (expected {}B, got {}B)",
+                        compressed_size, page_contents.size()));
+            }
+            return _source.advance(compressed_size).then([this, page_contents] {
+                return seastar::make_ready_future<std::optional<page>>(page{_latest_header.get(), page_contents});
+            });
+        });
+    });
+}
+
+std::basic_string_view<uint8_t>
+decompressor::operator()(std::basic_string_view<uint8_t> input, size_t decompressed_len) {
+    if (_codec == format::CompressionCodec::UNCOMPRESSED) {
+        return input;
+    } else {
+        if (decompressed_len > _buffer.size()) {
+            _buffer = buffer{decompressed_len};
+        }
+        switch (_codec) {
+        case format::CompressionCodec::SNAPPY:
+            compression::snappy_decompress(input.data(), input.size(), _buffer.data(), decompressed_len);
+            break;
+        case format::CompressionCodec::GZIP:
+            compression::zlib_decompress(input.data(), input.size(), _buffer.data(), decompressed_len);
+            break;
+        // TODO GZIP, LZO, BROTLI, LZ4, ZSTD
+        default:
+            throw parquet_exception::not_implemented(seastar::format("Unsupported compression ({})", _codec));
```

Comment:
nitpick: you're in seastar namespace (or at least you should be), so feel free to simply call `format()` - that's how it's usually done in our codebase.

Reply:
> you're in seastar namespace (or at least you should be)

Oh, should I? I have stayed out of `::seastar` intentionally. This project is hardly a part of Seastar the C++ async framework, and it's in this repository mostly for maintenance reasons, isn't it? But yes, come to think of it, if the headers are to live under `<seastar/*>`, then the namespace probably should reflect that. Should I rename `::parquet` to `::seastar::parquet` then?

> feel free to simply call format()

The advantage of keeping all names quantified is that ripping out the parquet library to a separate project (in case e.g. maintainers decide that this piece of code doesn't belong in Seastar after all) stays easy. So, are you sure about that? Should I make a pass over the entire thing and remove `seastar::` qualifications for readability?

Reply:
Pardon, missed your questions earlier:
1. For now we reside in seastar, so it's fine to rename ::parquet to ::seastar::parquet - even if we migrate to somewhere else, e.g. `apps/` directory, seastar namespace is still applicable. And if we migrate to, say, Scylla, then it will probably end up in yet another namespace anyway.
2. Since this library uses seastar future/promise primitives, I expect that if it was to be ripped out, it would end up either in `apps/` subdirectory (which is still part of seastar), or some other kind of subdirectory (again, part of seastar), or, perhaps Scylla. The thing with Scylla is that our internal coding style explicitly says that it's fine to assume `using namespace seastar` in the code base, since Scylla is heavily based on seastar. To sum up, I think it's a safe decision to drop explicit `seastar::` prefixes. Of course, they're also totally fine to stay, because they're correct - I'll leave the final choice to you, both ways are fine for me.